### DISCUSSION TITLE
Can risc0 be used to prove validity of non-EVM blocks?

### DISCUSSION THREAD

**Discussion by PaulRBerg** [risc0/risc0#892] [Q&A] [answered]
1. Can risc0 be used to prove the validity of non-EVM blocks? My intuition would say yes, but I just want to double-check this with you.
2. How difficult would this be? Can the [Rust starter](https://github.com/risc0/risc0/tree/main/templates/rust-starter) be used for this purpose?
3. Has anyone used risc0 to prove non-EVM blocks?

**Comment by mothran** [comment, marked as answer]
## 1)
The risc0 zkvm can be used for arbitrary compute and since our first tier supported language for the guest is Rust the kinda short answer is: "if you have rust code for, you can probably make it work within the ZKVM". Now there might be some dependencies that won't work but you can get a rough picture of our compatibility with: https://risc0.github.io/ghpages/dev/crate-validation/index.html

For example we have zeth / evm:
https://github.com/risc0/zeth
https://github.com/risc0/risc0/tree/main/examples/zkevm-demo (just processes single transactions)

But we also have examples for wasm:
https://github.com/risc0/risc0/tree/main/examples/wasm

## 2) 

Yes, I generally recommend starting with the rust starter template for rapid experiments of "hey will X work in the guest?". We have a good quick start guide here: https://dev.risczero.com/zkvm/quickstart

## 3)

Yes! I have worked with a few other L1s ISAs within the guest, mostly focused on projects that are rust focused and have had a lot of success just running different interpreters in the guest. There is obviously overhead but tracking cycle counts in the guest via:
https://docs.rs/risc0-zkvm/0.18.0/risc0_zkvm/guest/env/fn.get_cycle_count.html and https://docs.rs/risc0-zkvm/0.18.0/risc0_zkvm/guest/env/fn.log.html + `RISC0_DEV_MODE=true` will make it easy to rapidly get a approximate cycle cost of anything you doing in the guest.

**Reply by PaulRBerg** [reply]
It's rare that I receive such a detailed, thoughtful answer. Thank you so much.
